storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -117,5 +199,103 @@ func (s *peersSource) Read(
 
 	wg.Wait()
 
+	close(incrementalQueue)
+	incrementalWg.Wait()
+
+	if incremental {
+		// Now cache the incremental results
+		shards := make([]uint32, 0, len(shardsTimeRanges))
+		for shard := range shardsTimeRanges {
+			shards = append(shards, shard)
+		}
+
+		if err := blockRetriever.CacheShardIndices(shards); err != nil {
+			return nil, err
+		}
+	}
+
 	return result, nil
 }
+
+func (s *peersSource) incrementalFlush(
+	namespace ts.ID,
+	persistManager persist.Manager,
+	shard uint32,
+	shardRetrieverMgr block.DatabaseShardBlockRetrieverManager,
+	shardResult result.ShardResult,
+	tr xtime.Range,
+) error {
+	var (
+		numSeries      = len(shardResult.AllSeries())
+		flushedBlocks  = make([]incrementalFlushedBlock, 0, numSeries)
+		ropts          = s.opts.ResultOptions().RetentionOptions()
+		blockSize      = ropts.BlockSize()
+		shardRetriever = shardRetrieverMgr.ShardRetriever(shard)
+		tmpCtx         = context.NewContext()
+	)
+	for start := tr.Start; start.Before(tr.End); start = start.Add(blockSize) {
+		prepared, err := persistManager.Prepare(namespace, shard, start)
+		if err != nil {
+			return err
+		}
+
+		var blockErr error
+		flushedBlocks = flushedBlocks[:0]
+
+		for _, series := range shardResult.AllSeries() {
+			bl, ok := series.Blocks.BlockAt(start)
+			if !ok {
+				continue
+			}
+
+			tmpCtx.Reset()
+			stream, err := bl.Stream(tmpCtx)
+			if err != nil {
+				blockErr = err // Need to call prepared.Close, avoid return
+				break
+			}
+
+			segment, err := stream.Segment()
+			if err != nil {
+				blockErr = err // Need to call prepared.Close, avoid return
+				break
+			}
+
+			err = prepared.Persist(series.ID, segment, bl.Checksum())
+			tmpCtx.BlockingClose()
+			if err != nil {
+				blockErr = err // Need to call prepared.Close, avoid return
+				break
+			}
+
+			entry := incrementalFlushedBlock{
+				id:    series.ID,
+				block: bl,
+			}
+			flushedBlocks = append(flushedBlocks, entry)
+		}
+
+		// Always close before attempting to check if block error occurred,
+		// avoid using a defer here as this needs to be done for each inner loop
+		err = prepared.Close()
+		if blockErr != nil {
+			// A block error is worth bubbling up more than a close error
+			err = blockErr
+		}
+		if err != nil {
+			return err
+		}
+
+		// We can now make flushed blocks retrievable
+		for _, entry := range flushedBlocks {
```

Comment:
I wonder instead of accumulating all these flushed blocks and holding onto them until the flush succeeds, would it better to just reiterate over the result again so we don't need this `flushedBlocks` slice at all at the expense of slightly more CPU cycles?